src/raftstore/store/snap.rs

```diff
@@ -1302,15 +1302,19 @@ impl SnapManager {
         self.max_total_size
     }
 
-    pub fn register(&self, key: SnapKey, entry: SnapEntry) {
+    // For `SnapEntry::Sending` we need to support register multi times because
+    // Raft can send same snapshot to many peers (e.g. a follower and several learners).
+    // But for others, Only support register one time.
+    // TODO: make it work as a state machine, and change `Vec<SnapEntry>` to `SnapEntry`.
+    pub fn register(&self, key: SnapKey, entry: SnapEntry) -> Result<()> {
         debug!("register [key: {}, entry: {:?}]", key, entry);
-        let mut core = self.core.wl();
-        match core.registry.entry(key) {
+        match self.core.wl().registry.entry(key) {
             Entry::Occupied(mut e) => {
-                if e.get().contains(&entry) {
-                    warn!("{} is registered more than 1 time!!!", e.key());
-                    return;
+                if e.get().contains(&entry) && entry != SnapEntry::Sending {
+                    error!("{} is registered with {:?} multi times", e.key(), entry);
+                    return Err(box_err!("Register {:?} multi times", entry));
                 }
+                // Push `SnapEntry::Sending` multi times is allowed.
```

Comment:
Is it safe to register receiving when applying?

Reply:
It's safe because `new_for_receiving` will check the snapshot is exists or not. However the implemention is not simple or clear. I prefer to make the two states exlusive later.